datafusion/src/physical_plan/hash_aggregate.rs

```diff
@@ -339,6 +325,36 @@ pin_project! {
     }
 }
 
+fn hash_(group_values: &[ArrayRef]) -> Result<MutableBuffer<u64>> {
+    // compute the hashes
+    // todo: we should be able to use `MutableBuffer<u64>` to compute the hash and ^ them without
+    // allocating all the hashes before ^ them
+    let hashes = group_values
+        .iter()
+        .map(|x| {
+            let a = match x.data_type() {
+                DataType::Dictionary(_, d) => {
+                    // todo: think about how to perform this more efficiently
+                    // * first hash, then unpack
+                    // * do not unpack at all, and instead figure out a way to leverage dictionary-encoded.
+                    let unpacked = arrow2::compute::cast::cast(x.as_ref(), d)?;
+                    arrow2::compute::hash::hash(unpacked.as_ref())
+                }
+                _ => arrow2::compute::hash::hash(x.as_ref()),
+            };
+            Ok(a?)
+        })
+        .collect::<Result<Vec<_>>>()?;
+    let hash = MutableBuffer::<u64>::from(hashes[0].values());
+
+    Ok(hashes.iter().skip(1).fold(hash, |mut acc, x| {
+        acc.iter_mut()
+            .zip(x.values().iter())
+            .for_each(|(hash, other)| *hash ^= other);
```

Comment:
FYI this is a (extremely) bad way to combine hashes - I also learned this while improving the hash join!

E.g. see here: http://myeyesareblind.com/2017/02/06/Combine-hash-values/

Reply:
Simple example is that combining two same hashes always map to `0` - regardless of which value they have and in which order they appear.

Reply:
yeah, for me this was more like trying to get away from the `.value(i)`, which in arrow2 is very slow.

Reply:
In the hash join this `combine_hashes` function is used, which also could be used here:
```
fn combine_hashes(l: u64, r: u64) -> u64 {
    let hash = (17 * 37u64).wrapping_add(l);
    hash.wrapping_mul(37).wrapping_add(r)
}
```

e.g.  something like: `*hash ^= combine_hashes(hash, other)`

Reply:
Thanks @Dandandan , this is now migrated; it uses the same that `hash_join` does.